crates/astria-gossipnet/src/network_stream.rs

```diff
@@ -0,0 +1,273 @@
+use std::{
+    pin::Pin,
+    str::FromStr,
+    task::{
+        Context,
+        Poll,
+    },
+};
+
+use color_eyre::eyre::{
+    eyre,
+    Result,
+};
+use futures::StreamExt;
+use libp2p::{
+    gossipsub::{
+        self,
+        Message,
+        TopicHash,
+    },
+    kad::{
+        Addresses,
+        KademliaEvent,
+        QueryResult,
+    },
+    mdns,
+    swarm::SwarmEvent,
+    Multiaddr,
+    PeerId,
+};
+use tracing::debug;
+
+use crate::network::{
+    GossipnetBehaviourEvent,
+    Network,
+};
+
+#[derive(Debug)]
+pub enum Event {
+    // Swarm events
+    NewListenAddr(Multiaddr),
+
+    // mDNS events
+    MdnsPeersConnected(Vec<PeerId>),
+    MdnsPeersDisconnected(Vec<PeerId>),
+
+    // Gossipsub events
+    GossipsubPeerConnected(PeerId),
+    GossipsubPeerSubscribed(PeerId, TopicHash),
+    GossipsubMessage(Message),
+
+    // Kademlia events
+    RoutingUpdated(PeerId, Addresses),
+}
+
+impl futures::Stream for Network {
+    type Item = Result<Event>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        while let Poll::Ready(maybe_event) = self.swarm.poll_next_unpin(cx) {
+            let Some(event) = maybe_event else {
+                self.terminated = true;
```

Comment:
Is this ever set to `false` actually?

Reply:
no, it's only initially false by default, but never set to false again after being true.

Reply:
I'll remove this actually as it doesn't do anything atm